bob/src/cluster/quorum.rs

```diff
@@ -32,33 +32,34 @@ impl Quorum {
 
     async fn put_at_least(&self, key: BobKey, data: &BobData) -> Result<(), Error> {
         let mut local_put_ok = 0_usize;
-        let at_least = self.quorum;
+        let mut at_least = self.quorum;
         let mut failed_nodes = Vec::new();
-        let (vdisk_id, disk_path) = self.mapper.get_operation(key);
-        let (tasks, errors) =
-            if let Some(disk_path) = disk_path {
-                debug!("PUT[{}] ~~~PUT TO {} REMOTE NODES AND LOCAL NODE~~~", key, at_least - 1);
-                let (remote_put, local_put) = tokio::join!(
-                    self.put_remote_nodes(key, data, at_least - 1),
-                    put_local_node(&self.backend, key, data, vdisk_id, disk_path),
-                );
-                let (mut remote_tasks, mut errors) = remote_put;
-                if let Err(e) = local_put {
-                    error!("{}", e);
-                    failed_nodes.push(self.mapper.local_node_name().clone());
-                    debug!("PUT[{}] local failed, put another remote", key);
-                    errors.extend(finish_at_least_handles(&mut remote_tasks, 1).await.into_iter());
-                } else {
-                    local_put_ok += 1;
-                    debug!("PUT[{}] local node put successful", key);
+        let (vdisk_id, disk_paths) = self.mapper.get_operation(key);
+        let affected_replicas_by_node = self.mapper.get_replicas_count_by_node(key);
+        let (tasks, errors) = 
+            if let Some(paths) = disk_paths {
+                let paths_len = paths.len();
+                debug!("PUT[{}] ~~~PUT TO {} REMOTE NODES AND LOCAL NODE~~~", key, at_least - paths_len);
+                let ((tasks, errors), local_puts) = tokio::join!(
+                    self.put_remote_nodes(key, data, at_least, &affected_replicas_by_node),
+                    put_local_node_all(&self.backend, key, data, vdisk_id, paths));
+                if local_puts != paths_len {
+                    failed_nodes.push(self.mapper.local_node_name().to_owned());
                 }
-                (remote_tasks, errors)
+                local_put_ok += local_puts;
+                at_least -= local_puts;
+                (tasks, errors)
             } else {
-                debug!("PUT[{}] ~~~PUT TO {} REMOTE NODES~~~", key, at_least);
-                self.put_remote_nodes(key, data, at_least).await
+                debug!("PUT[{}] ~~~PUT TO REMOTE NODES~~~", key);
+                self.put_remote_nodes(key, data, at_least, &affected_replicas_by_node).await
             };
-        let all_count = self.mapper.get_target_nodes_for_key(key).len();
-        let remote_ok_count = all_count - errors.len() - tasks.len() - local_put_ok;
+        debug!("PUT[{}] need at least {} additional puts", key, at_least);
+        let target_nodes = self.mapper.get_target_nodes_for_key(key);
+        let all_count: usize = target_nodes.iter()
+            .map(|n| *affected_replicas_by_node.get(n.name()).unwrap_or(&1))
```

Comment:
All node names returned by `get_target_nodes_for_key(key)` should be presented inside `affected_replicas_by_node`, so it is better to use `expect("...")` instead of `unwrap_or(..)`